src/protocols/secure_channel/SimpleSessionResumptionStorage.cpp

```diff
@@ -0,0 +1,273 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP CASE Session object that provides
+ *      APIs for constructing a secure session using a certificate from the device's
+ *      operational credentials.
+ */
+
+#include <protocols/secure_channel/SimpleSessionResumptionStorage.h>
+
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+
+namespace chip {
+
+constexpr TLV::Tag SimpleSessionResumptionStorage::kIndexContentTag;
+constexpr TLV::Tag SimpleSessionResumptionStorage::kFabricIndexTag;
+constexpr TLV::Tag SimpleSessionResumptionStorage::kPeerNodeIdTag;
+constexpr TLV::Tag SimpleSessionResumptionStorage::kResumptionIdTag;
+constexpr TLV::Tag SimpleSessionResumptionStorage::kSharedSecretTag;
+constexpr TLV::Tag SimpleSessionResumptionStorage::kCATTag;
+
+const char * SimpleSessionResumptionStorage::StorageKey(DefaultStorageKeyAllocator & keyAlloc, const ScopedNodeId & node)
+{
+    return keyAlloc.FabricSession(node.GetFabricIndex(), node.GetNodeId());
+}
+
+const char * SimpleSessionResumptionStorage::StorageKey(DefaultStorageKeyAllocator & keyAlloc, ConstResumptionIdView resumptionId)
+{
+    char resumptionIdBase64[BASE64_ENCODED_LEN(resumptionId.size()) + 1];
+    auto len                = Base64Encode(resumptionId.data(), resumptionId.size(), resumptionIdBase64);
+    resumptionIdBase64[len] = '\0';
+    return keyAlloc.SessionResumption(resumptionIdBase64);
+}
+
+CHIP_ERROR SimpleSessionResumptionStorage::SaveIndex(const SessionIndex & index)
+{
+    uint8_t buf[MaxIndexSize()];
+    TLV::TLVWriter writer;
+    writer.Init(buf);
+
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(writer.StartContainer(kIndexContentTag, TLV::kTLVType_Array, arrayType));
+
+    for (size_t i = index.mSize; i < index.mSize; ++i)
+    {
+        TLV::TLVType innerType;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, innerType));
+        ReturnErrorOnFailure(writer.Put(kFabricIndexTag, index.mNodes[i].GetFabricIndex()));
+        ReturnErrorOnFailure(writer.Put(kPeerNodeIdTag, index.mNodes[i].GetNodeId()));
+        ReturnErrorOnFailure(writer.EndContainer(innerType));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(arrayType));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    DefaultStorageKeyAllocator keyAlloc;
+    ReturnErrorOnFailure(mStorage->SyncSetKeyValue(keyAlloc.SessionResumptionIndex(), buf, static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSessionResumptionStorage::LoadIndex(SessionIndex & index)
+{
+    uint8_t buf[MaxIndexSize()];
+    uint16_t len = static_cast<uint16_t>(MaxStateSize());
```

Comment:
This is wrong.  `len` shoud be set to `MaxIndexSize()`, or better yet `ArraySize(buf)`.  Why did the tests miss this?